File: IP/3rd/sampleprogram/11-3-9.cpp
#define _CRT_SECURE_NO_WARNINGS
#define _USE_MATH_DEFINES
#include <iostream>
#include <cmath>
#include <opencv2/opencv.hpp>
using namespace std;
using namespace cv;
string win_src = "src";
string win_dst = "dst";

int main()
{
  string file_src = "src_depth_ball.png"; // 入力画像のファイル名
  string file_dst = "dst.png"; // 出力画像のファイル名
  // Mat img_src = imread(file_src, 1); // 入力画像（カラー）の読み込み
  Mat img_src = imread(file_src, 0); // 入力画像（グレースケール）の読み込み

  Mat img_dst;
  if(!img_src.data) {
    cout << "error" << endl;
    return -1;
  }

  // ウインドウ生成
  namedWindow(win_src, WINDOW_AUTOSIZE);
  namedWindow(win_dst, WINDOW_AUTOSIZE);

  // セグメンテーション
  int depth_min = 206;
  int depth_max = 210;
  inRange(img_src, depth_min, depth_max, img_dst);

  // ノイズ除去（収縮・膨張）
  erode(img_dst, img_dst, Mat(), Point(-1, -1), 3);
  dilate(img_dst, img_dst, Mat(), Point(-1, -1), 3);

  // ラベリング
  Mat img_lab;
  int nlabel = connectedComponents(img_dst, img_lab);
  for(int i = 1; i < nlabel; i++) {
    Mat img_dst;
    compare(img_lab, i, img_dst, CMP_EQ); // ラベルiを抜き出し
    cout << i << " / " << (nlabel - 1) << endl;
    imwrite("dst" + to_string(i) + ".png", img_dst);
    imshow(win_src, img_src);
    imshow(win_dst, img_dst);
    waitKey(0);
  }

  return 0;
}

File: OpenCVによる画像処理入門/第3版/第2刷/sampleprogram/11-3-9.py
import cv2
import math
import numpy as np

file_src = 'src_depth_ball.png'
file_dst = 'dst.png'

img_src = cv2.imread(file_src, cv2.IMREAD_GRAYSCALE)

cv2.namedWindow('src')
cv2.namedWindow('dst')

depth_min = 206
depth_max = 210
img_dst = cv2.inRange(img_src, depth_min, depth_max)  # セグメンテーション

# ノイズ除去（収縮・膨張）
kernel = np.ones((3, 3), np.uint8)
img_dst = cv2.erode(img_dst, kernel, iterations=2)
img_dst = cv2.dilate(img_dst, kernel, iterations=2)

# ラベリング
nlabel, img_lab = cv2.connectedComponents(img_dst)
for i in range(1, nlabel, 1):
  img_dst = cv2.compare(img_lab, i, cv2.CMP_EQ)  # ラベルiを抜き出し
  print(i, ' / ', (nlabel - 1))
  cv2.imwrite('dst' + str(i) + '.png', img_dst)
  cv2.imshow('src', img_src)
  cv2.imshow('dst', img_dst)
  cv2.waitKey(0)

cv2.destroyAllWindows()
